flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.minicluster.RpcServiceSharing;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+/**
+ * Starts a {@link MiniCluster} for every submitted job.
+ * This class guarantees to tear down the MiniCluster in case of normal or exceptional job completion.
+ * */
+public class PerJobMiniCluster {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PerJobMiniCluster.class);
+
+	private final Configuration configuration;
+	private final Function<MiniClusterConfiguration, MiniCluster> miniClusterFactory;
+
+	public PerJobMiniCluster() {
+		this(new Configuration());
+	}
+
+	public PerJobMiniCluster(Configuration configuration) {
+		this(configuration, MiniCluster::new);
+	}
+
+	public PerJobMiniCluster(Configuration configuration, Function<MiniClusterConfiguration, MiniCluster> miniClusterFactory) {
+		this.configuration = configuration;
+		this.miniClusterFactory = miniClusterFactory;
+	}
+
+	/**
+	 * Starts a {@link MiniCluster} and submits a job.
+	 */
+	public CompletableFuture<JobClient> submitJob(JobGraph jobGraph) throws Exception {
+		MiniClusterConfiguration miniClusterConfig = getMiniClusterConfig(jobGraph.getMaximumParallelism());
+		MiniCluster miniCluster = miniClusterFactory.apply(miniClusterConfig);
+		miniCluster.start();
+
+		return miniCluster
+			.submitJob(jobGraph)
+			.handle((result, throwable) -> {
+				if (throwable != null) {
+					try {
+						throw new RuntimeException("Error submitting job to MiniCluster.", throwable);
+					} finally {
+						shutDownCluster(miniCluster);
+					}
+				}
+				return new PerJobMiniClusterJobClient(result.getJobID(), miniCluster);
+			});
+	}
+
+	private MiniClusterConfiguration getMiniClusterConfig(int maximumParallelism) {
+		Configuration configuration = this.configuration.clone();
+
+		if (!configuration.contains(RestOptions.BIND_PORT)) {
+			configuration.setString(RestOptions.BIND_PORT, "0");
+		}
+
+		int numTaskManagers = configuration.getInteger(
+			ConfigConstants.LOCAL_NUMBER_TASK_MANAGER,
+			ConfigConstants.DEFAULT_LOCAL_NUMBER_TASK_MANAGER);
+
+		// we have to use the maximum parallelism as a default here, otherwise streaming pipelines would not run
+		int numSlotsPerTaskManager = configuration.getInteger(TaskManagerOptions.NUM_TASK_SLOTS, maximumParallelism);
+
+		return new MiniClusterConfiguration.Builder()
+			.setConfiguration(configuration)
+			.setNumTaskManagers(numTaskManagers)
+			.setRpcServiceSharing(RpcServiceSharing.SHARED)
+			.setNumSlotsPerTaskManager(numSlotsPerTaskManager)
+			.build();
+	}
+
+	private static void shutDownCluster(MiniCluster miniCluster) {
+		try {
+			miniCluster.closeAsync().get();
```

Comment:
Blocking calls should be avoided if possible. The problem is that you have to make sure that this method won't be called from a `Thread` belonging to `miniCluster`. Otherwise this can deadlock because the `ThreadPool` cannot shut down if one of its thread is waiting here.